File: AnimationDemo/app/src/main/java/com/geng/animationdemo/MainActivity.java
package com.geng.animationdemo;

import android.app.FragmentManager;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.widget.Button;

public class MainActivity extends AppCompatActivity implements View.OnClickListener {

    private Button accelerateBt;
    private Button decelerateBt;
    private Button otherBt;
    private FragmentManager fragmentManager;
    private AccelerateFragment accelerateFragment;
    private DecelerateFragment decelerateFragment;
    private OtherFragment otherFragment;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        initView();
        initData();
        initListener();
    }

    private void initView() {
        this.otherBt = (Button) findViewById(R.id.otherBt);
        this.decelerateBt = (Button) findViewById(R.id.decelerateBt);
        this.accelerateBt = (Button) findViewById(R.id.accelerateBt);
    }

    private void initData() {
        fragmentManager = getFragmentManager();
        accelerateFragment = new AccelerateFragment();
        decelerateFragment = new DecelerateFragment();
        otherFragment = new OtherFragment();

        fragmentManager
                .beginTransaction()
                .replace(R.id.container, accelerateFragment)
                .commit();
    }

    private void initListener() {
       accelerateBt.setOnClickListener(this);
       decelerateBt.setOnClickListener(this);
       otherBt.setOnClickListener(this);
    }

    @Override
    public void onClick(View view) {
        switch (view.getId()) {
            case R.id.accelerateBt:
                if (accelerateFragment.isAdded()) {
                    fragmentManager
                            .beginTransaction()
                            .hide(decelerateFragment)
                            .hide(otherFragment)
                            .show(accelerateFragment)
                            .commit();
                } else {
                    fragmentManager
                            .beginTransaction()
                            .add(R.id.container, accelerateFragment)
                            .commit();
                }
                break;
            case R.id.decelerateBt:
                if (decelerateFragment.isAdded()) {
                    fragmentManager
                            .beginTransaction()
                            .hide(accelerateFragment)
                            .hide(otherFragment)
                            .show(decelerateFragment)
                            .commit();
                } else {
                    fragmentManager
                            .beginTransaction()
                            .add(R.id.container, decelerateFragment)
                            .commit();
                }
                break;
            case R.id.otherBt:
                if (otherFragment.isAdded()) {
                    fragmentManager
                            .beginTransaction()
                            .hide(decelerateFragment)
                            .hide(accelerateFragment)
                            .show(otherFragment)
                            .commit();
                } else {
                    fragmentManager
                            .beginTransaction()
                            .add(R.id.container, otherFragment)
                            .commit();
                }
                break;
        }
    }
}

File: StatusBarDemo/app/src/main/java/com/gjr/statusbardemo/MainActivity.kt
package com.gjr.statusbardemo

import android.graphics.Color
import android.os.Build
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.view.WindowManager
import android.widget.FrameLayout
import android.widget.Toast
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        // Application主题： @style/Theme.AppCompat.Light.NoActionBar


        val version = Build.VERSION.SDK_INT;
        val flag = false


        if (version >= Build.VERSION_CODES.KITKAT
                && version < Build.VERSION_CODES.LOLLIPOP) {
            if (flag) {
                // 顶部的控件显示在状态栏下面，不被状态栏遮挡
                (findViewById(android.R.id.content) as ViewGroup).getChildAt(0).fitsSystemWindows = true
                // 设置状态栏透明色
                window.addFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS)

                // 思路： 在decorView中添加个view(和状态栏一样高)，来改变状态栏样式
                val statusView = View(this)
                val statusBarHeight = getStatusBarHeight()
                val lp = FrameLayout.LayoutParams(FrameLayout.LayoutParams.MATCH_PARENT, statusBarHeight)
                lp.gravity = Gravity.TOP;
                statusView.layoutParams = lp;
                statusView.setBackgroundColor(Color.parseColor("#FF4081"))//#303F9F   #FF4081
                (window.decorView as ViewGroup).addView(statusView)
            } else {
                // 设置状态栏透明色
                window.addFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS)
                window.decorView.systemUiVisibility = View.SYSTEM_UI_FLAG_LAYOUT_STABLE or View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN

                val mChildView = (findViewById(android.R.id.content) as ViewGroup).getChildAt(0)
                if (mChildView != null) {
                    mChildView.setPadding(
                            mChildView.paddingLeft,
                            0,
                            mChildView.paddingRight,
                            mChildView.paddingBottom
                    )
                }
            }
        } else if (version >= Build.VERSION_CODES.LOLLIPOP) {
            if (flag) {
                window.statusBarColor = Color.parseColor("#FF4081")
            } else {
                // 全屏显示
                window.clearFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS)
                window.statusBarColor = Color.TRANSPARENT
                window.decorView.systemUiVisibility = View.SYSTEM_UI_FLAG_LAYOUT_STABLE or View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
            }
        }


        // 在application 主题中设置 fitsSystemWindows = true, toast 显示会有问题，压扁
        btn.setOnClickListener { Toast.makeText(this, "哈哈哈哈", Toast.LENGTH_LONG).show() }
    }

    private fun getStatusBarHeight(): Int {
        val resId = resources.getIdentifier("status_bar_height", "dimen", "android")
        if (resId > 0) {
            return resources.getDimensionPixelSize(resId)
        }
        return 0
    }
}
